pkg/ddc/jindofsx/sync_runtime.go

```diff
@@ -0,0 +1,236 @@
+/*
+Copyright 2022 The Fluid Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jindofsx
+
+import (
+	"context"
+	"reflect"
+
+	datav1alpha1 "github.com/fluid-cloudnative/fluid/api/v1alpha1"
+	"github.com/fluid-cloudnative/fluid/pkg/ctrl"
+	fluiderrs "github.com/fluid-cloudnative/fluid/pkg/errors"
+	cruntime "github.com/fluid-cloudnative/fluid/pkg/runtime"
+	"github.com/fluid-cloudnative/fluid/pkg/utils"
+	"github.com/fluid-cloudnative/fluid/pkg/utils/kubeclient"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+)
+
+// SyncRuntime syncs the runtime spec
+func (e *JindoFSxEngine) SyncRuntime(ctx cruntime.ReconcileRequestContext) (changed bool, err error) {
+	runtime, err := e.getRuntime()
+	if err != nil {
+		return changed, err
+	}
+
+	// 1. sync master
+	masterChanged, err := e.syncMasterSpec(ctx, runtime)
+	if err != nil {
+		return
+	}
+	if masterChanged {
+		e.Log.Info("Master Spec is updated", "name", ctx.Name, "namespace", ctx.Namespace)
+		return masterChanged, err
+	}
+
+	// 2. sync workers
+	workerChanged, err := e.syncWorkerSpec(ctx, runtime)
+	if err != nil {
+		return
+	}
+	if workerChanged {
+		e.Log.Info("Worker Spec is updated", "name", ctx.Name, "namespace", ctx.Namespace)
+		return workerChanged, err
+	}
+
+	// 3. sync fuse
+	fuseChanged, err := e.syncFuseSpec(ctx, runtime)
+	if err != nil {
+		return
+	}
+	if fuseChanged {
+		e.Log.Info("Fuse Spec is updated", "name", ctx.Name, "namespace", ctx.Namespace)
+		return fuseChanged, err
+	}
+
+	return
+}
+
+func (e *JindoFSxEngine) syncMasterSpec(ctx cruntime.ReconcileRequestContext, runtime *datav1alpha1.JindoRuntime) (changed bool, err error) {
+	e.Log.V(1).Info("syncMasterSpec")
+	if runtime.Spec.Master.Disabled {
+		return
+	}
+
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		master, err := kubeclient.GetStatefulSet(e.Client, e.getMasterName(), e.namespace)
+		if err != nil {
+			return err
+		}
+
+		if len(runtime.Spec.Master.Resources.Limits) == 0 && len(runtime.Spec.Master.Resources.Requests) == 0 {
+			e.Log.V(1).Info("The resource requirement is not set, skip")
+			return nil
+		}
+
+		masterToUpdate := master.DeepCopy()
+		if len(masterToUpdate.Spec.Template.Spec.Containers) == 1 {
+			masterResources := utils.TranformResourcesWithTieredStore(runtime.Spec.Master.Resources,
```

Comment:
Does jindofsx master also needs `TransformResourceWithTieredStore`? Jindofsx master does not have tieredstore so maybe it doesn't need it?

Reply:
Because master's memory also relates to the tieredStore, the meta store's size depends on the data scalablity to cache.